### PR TITLE
Nvfuser to copy decomp to prim

### PR DESCRIPTION
**Pull request by jjsjann123**
Conditional decomposing aten::_to_copy to nvprim::convert_element_type to allow fusion with type casting, which is introduced during type promotion phase at torch decomposition.

### CODE REVIEW

**Comment by IvanYashchuk** [comment]
A less intrusive and more primtorch-general approach would be:
1.  Add `torch.Tensor.to: torch._refs.to,` entry in `torch_to_refs_map` in context.py
2. Implement `torch._refs.to` function that would call `prims.convert_element_type` for dtype arg and fallback to `torch.Tensor.to` otherwise, the signature of the `to` function is here: https://github.com/pytorch/pytorch/blob/5a1f6d50a9a4059315c5530703e5dc9e38229715/torch/overrides.py#L1318

**Comment by Chillee** [comment]
@IvanYashchuk I think that would work for decomps, but I assume there are also other places that _to_copy shows up in the trace that simply interposing with a __torch_function__ mode wouldn’t capture.

IMO the most “primtorch native” way would be to simply write a ref/decomp for _to_copy (maybe it could just be mapped to torch.to?)

**Comment by IvanYashchuk** [comment]
I disagree, in primtorch project the primary goal is to provide reference implementations for Python user-facing interface, not for internal ATen CompositeExplicitAutograd functions.

I don't see how `aten::_to_copy` would appear in the graph without explicit call to `torch.ops.aten._to_copy` from Python script after my suggestion is implemented.

> IMO the most “primtorch native” way would be to simply write a ref/decomp for _to_copy (maybe it could just be mapped to torch.to?)

Decompositions for internal ATen CompositeExplicitAutograd functions should go to `torch/_decomp/decompositions.py` to be used by functorch and other related projects.

**Comment by Chillee** [comment]
> I don't see how aten::_to_copy would appear in the graph without explicit call to torch.ops.aten._to_copy from Python script after my suggestion is implemented.

Well, since we're tracing through C++, anywhere that we call `to` in C++. One common case is with AMP.

```
def f(x):
    x = x * 2
    x = x @ x
    x = x * 2
    return x


with torch.autocast('cpu'):
    result = make_fx(f)(torch.randn(5, 5, dtype=torch.float16))
    print(result.code)
```

results in
```
def forward(self, x_1):
    mul_tensor = torch.ops.aten.mul.Tensor(x_1, 2);  x_1 = None
    _to_copy_default = torch.ops.aten._to_copy.default(mul_tensor, dtype = torch.bfloat16)
    _to_copy_default_1 = torch.ops.aten._to_copy.default(mul_tensor, dtype = torch.bfloat16);  mul_tensor = None
    mm_default = torch.ops.aten.mm.default(_to_copy_default_1, _to_copy_default);  _to_copy_default_1 = _to_copy_default = None
    mul_tensor_1 = torch.ops.aten.mul.Tensor(mm_default, 2);  mm_default = None
    return mul_tensor_1
```

Presumably we want to lower those type promotions to NVFuser as well.

> Decompositions for internal ATen CompositeExplicitAutograd functions should go to torch/_decomp/decompositions.py to be used by functorch and other related projects.

I thought we were making those refs as well, hence https://github.com/pytorch/pytorch/pull/79413.

**Comment by IvanYashchuk** [comment]
Thank you for the autocast example. I forgot that autocast was implemented in C++. But we're not testing anything with autocast enabled anyway, so I expect there are other failures for this and other things that aren't tested.

Currently, we're concerned with being able to use decompositions from `torch/_decomp/decompositions.py` that are implemented in Python, and the test implemented in this PR passes with my suggestion. It's of course better to have a full implementation of the `to` function, but for now, it can fallback to `torch.ops.aten.to` for all other arguments other than `dtype`.


> I thought we were making those refs as well, hence https://github.com/pytorch/pytorch/pull/79413.

`torch.native_layer_norm` is a part of the public PyTorch interface since it's available from "torch.".

**Comment by jjsjann123** [comment]
hmmm.. trying to get torch._refs.to but it's getting messy. :cry: 

Marking progress, listing a few tests that has failed on me with a few of my local changes failures:
```
python test/test_proxy_tensor.py -k test_make_fx_symbolic_exhaustive_nn_functional_logsigmoid
python test/test_ops.py -k test_refs_are_in_python_ref_db_op
python test/test_ops.py -k test_refs_are_in_decomp_table_op
```

**Comment by ngimel** [comment]
_to_copy decomposition now available #84108, you can switch to using it. If memory_format arg is supplied it probably won't do you any good, but for usages in type promotion wrappers that should be enough. To better support memory_format we'd need convert_memory_format prim, PRs accepted.

File: torch/_refs/__init__.py [hunk 1]
```diff
@@ -1654,6 +1655,27 @@ def item(a: TensorLikeType) -> NumberType:
     return number_type(prims.item(a))
 
 
+@register_decomposition(torch.ops.aten.to)
```
Comment (IvanYashchuk) [inline torch/_refs/__init__.py:1658]: Let's not do this. If some test that checks that all refs are registered is failing we can add this function to the skip list.

File: torch/_refs/__init__.py [hunk 2]
```diff
@@ -1654,6 +1655,27 @@ def item(a: TensorLikeType) -> NumberType:
...
+    non_blocking: bool = False,
+    copy: bool = False,
+    memory_format: torch.memory_format = torch.preserve_format,
+) -> TensorLikeType:
+    if (
+        (copy is True or dtype != a.dtype)
+        and memory_format == torch.preserve_format
+        and non_blocking is False
+    ):
+        return torch.ops.nvprims.convert_element_type(a, dtype)
```
Comment (IvanYashchuk) [inline torch/_refs/__init__.py:1671]: _refs module should use plain prims
```suggestion
        return torch.ops.prims.convert_element_type(a, dtype)
```

**Comment by IvanYashchuk** [comment]
`python test/test_ops.py -k test_refs_are_in_python_ref_db_op`
It complains about a reference being implemented without a corresponding OpInfo entry. No one has added one before.

For `python test/test_ops.py -k test_refs_are_in_decomp_table_op` we can skip the test by adding `'_refs.to'` to `not_in_decomp_table` inside `test_ops.py`.

Also `python test/test_proxy_tensor.py -k test_make_fx_symbolic_exhaustive_nn_functional_logsigmoid` should be passing if we don't register this `to` implementation for the decomposition table, because make_fx_symbolic doesn't like `empty` call.

**Comment by jjsjann123** [comment]
patched failing CI, switched to upstream _to_copy decomp as @ngimel suggested. Local tests passed for previously failing CI.

Let's see how CI goes this time.

Meanwhile, it should be good for a second round of review.

**Comment by jjsjann123** [comment]
:tada:  Looks like CI is clean now~~~

**Review by ngimel — APPROVED** [review]

File: test/test_prims.py [hunk 3]
```diff
@@ -197,7 +197,7 @@ def func(a):
             return torch.digamma(a)
 
         with TorchRefsNvfuserCapabilityMode():
-            gm = make_fx(func)(a)
+            gm = make_fx(func, decomposition_table=torch._prims.context.nvfuser_decomp_table())(a)
```
[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply test/test_prims.py:200]: This test passes without this change.

File: test/test_prims.py [hunk 4]
```diff
@@ -509,5 +509,42 @@ def test_constant_pad_nd_memory_format(self, device, dtype):
 instantiate_device_type_tests(TestRefs, globals())
 
 
+class TestDecomp(TestCase):
+    @onlyCUDA
+    @skipCUDAIfRocm
+    @dtypes(torch.float16, torch.float32)
+    def test_decomposition_type_promotion_nvprim_amp(self, device, dtype):
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply test/test_prims.py:516]: Sounds about right. Are you suggesting that we should break it into multiple smaller tests instead?

**Pull request closed without merging**